### DISCUSSION TITLE
TFT35v3 B1 and BTT SKR 1.4 (BIQU B1) Feedback

### DISCUSSION THREAD

**Discussion by pixeldoc2000** [teeminus/NoTouchScreenFirmware#43] [General]
This works on my BIQU B1 Printer with SKR 1.4 Board and TFT35v3 B1 Display with NoTouchScreenFirmware v1.3.
Klipper: v0.9.1-647-gbe122411

```
[board_pins]
aliases:
    # EXP1 header
    EXP1_1=P1.30, EXP1_3=P1.18, EXP1_5=P1.20, EXP1_7=P1.22, EXP1_9=<GND>,
    EXP1_2=P0.28, EXP1_4=P1.19, EXP1_6=P1.21, EXP1_8=P1.23, EXP1_10=<5V>,
    # EXP2 header
    EXP2_1=P0.17, EXP2_3=P3.26, EXP2_5=P3.25, EXP2_7=P1.31, EXP2_9=<GND>,
    EXP2_2=P0.15, EXP2_4=P0.16, EXP2_6=P0.18, EXP2_8=<RST>, EXP2_10=<NC>
    # Pins EXP2_1, EXP2_6, EXP2_2 are also MISO, MOSI, SCK of bus "ssp0"

[display]
lcd_type: emulated_st7920
en_pin: EXP1_4
spi_software_sclk_pin: EXP1_5
spi_software_mosi_pin: EXP1_3
spi_software_miso_pin: EXP1_6
encoder_pins: ^EXP2_5, ^EXP2_3
click_pin: ^!EXP1_2
```
EDIT: Display still freezes at some point while printing.

**Comment by teeminus** [comment]
Thanks for the feedback 👍

**Comment by pixeldoc2000** [comment]
Any Idea why the Display still freezes (does not Update and Knob does nothing) at some point while printing or should i open an issus for this?

**Comment by teeminus** [comment]
You could open an issue for that but I have no idea why this happens... Or even if the issue is caused by the display or by Klipper (which would include the mainboard as well). Have you checked the Klipper logs for errors?

I never had any freezing issues with my TFT35v3, even when the printer works all day long.

**Comment by pixeldoc2000** [comment]
Hm, 'klippy.log' has no error that I can spot.

The weird thing is, it does only happen on about 9/10 prints. Most often it happens when starting a print on the "G28 Home all axes".

Example of a "freeze":
![photo_2021-08-03_23-30-21](https://user-images.githubusercontent.com/376715/128089049-b36c2bb1-dcef-486e-aa0e-1a5590e95458.jpg)

Then I noticed artifacts on the Display after the Print was finished and the Screen did not freeze:
![photo_2021-08-03_23-49-59](https://user-images.githubusercontent.com/376715/128091019-2c14db61-5b53-4510-a47d-46105cea882b.jpg)

I am going to run some Tests with the Debug Firmware now.

Maybe the I need to use another `spi_software_miso_pin`?

**Reply by pixeldoc2000** [reply]
The Screen keeps freezing, even with another unused `spi_software_miso_pin`.

When the Screen freezes / does not update I can see incoming data with the Debug FW on the top left side of the screen as described in the WIKI at Troubleshooting. The top right Side does not update.

**Reply by pixeldoc2000** [reply]
Weird, it looks like with the current klipper version (v0.9.1-667-g31761500) and the stock display driver `lcd_type: st7920` it does not freeze and there are no artifacts.

I am sure it was the case with an older klipper version. That is why I migrated to the new display driver...

**Reply by teeminus** [reply]
Very strange. I initially wrote the new display driver because the stock one was not working for all displays. So far I never had any issues with the emulated_st7920 driver but that's for my setup. Maybe there are side effects when using the software SPI, maybe the speed of the bus is to low. You could try changing the speed of the software spi for the emulated driver as well.

**Reply by pixeldoc2000** [reply]
Tried every possible `spi_speed` setting and different pins with `lcd_type: emulated_st7920`, but the Display still freezes at some point while Homing or Printing.

Today I've gone back to `lcd_type: st7920` again and had two successful prints without the Display freezing, what do you know...

Still using `NoTouchFw v1.3 Debug`.

Klipper: v0.10.0-84-g7d95a004
```
[board_pins]
aliases:
    # EXP1 header - LCD
    EXP1_1=P1.30, EXP1_3=P1.18, EXP1_5=P1.20, EXP1_7=P1.22, EXP1_9=<GND>,
    EXP1_2=P0.28, EXP1_4=P1.19, EXP1_6=P1.21, EXP1_8=P1.23, EXP1_10=<5V>,
    # EXP2 header
    EXP2_1=P0.17, EXP2_3=P3.26, EXP2_5=P3.25, EXP2_7=P1.31, EXP2_9=<GND>,
    EXP2_2=P0.15, EXP2_4=P0.16, EXP2_6=P0.18, EXP2_8=<RST>, EXP2_10=<NC>
    # Pins EXP2_1, EXP2_6, EXP2_2 are also MISO, MOSI, SCK of bus "ssp0"

[display]
lcd_type: st7920
cs_pin: EXP1_4
sclk_pin: EXP1_5
sid_pin: EXP1_3
#encoder_pins: ^EXP2_3, ^EXP2_5
encoder_pins: ^EXP2_5, ^EXP2_3
click_pin: ^!EXP1_2
```

**Reply by teeminus** [reply]
That's great! If the stock driver works for you, stick to it.

I still think the issue is caused by the mainboard not being fast enough writing the data to the display which causes the malfunction. But other than that I am out of ideas.

**Comment by pixeldoc2000** [comment]
Same issus #47 with BIQU B1

Tried different `spi_speed` settings, makes no differents even with current klipper version.

**Comment by pixeldoc2000** [comment]
Possible Fix: https://github.com/Klipper3d/klipper/issues/2231#issuecomment-975419955

Different MISO PIN:
```
[display]
spi_speed: 1000000
spi_software_miso_pin: EXP2_1
```

I will test it and report back.
